File: samples/nim/trie/src/trie.nim
type
  TrieNode* = ref object
    children: array[26, TrieNode]
    isEndOfWord: bool

  Trie* = object
    root: TrieNode


proc newTrieNode(): TrieNode =
  TrieNode(
    isEndOfWord: false
  )


proc newTrie*(): Trie =
  result.root = newTrieNode()


proc getRoot*(t: Trie): TrieNode =
  t.root


proc insert*(t: Trie, word: string) =
  var current = t.root

  for c in word:
    let index = ord(c) - ord('a')

    if current.children[index] == nil:
      current.children[index] = newTrieNode()

    current = current.children[index]

  current.isEndOfWord = true


proc search*(t: Trie, word: string): bool =
  var current = t.root

  for c in word:
    let index = ord(c) - ord('a')

    if current.children[index] == nil:
      return false

    current = current.children[index]

  return current.isEndOfWord


proc startsWith*(t: Trie, prefix: string): bool =
  var current = t.root

  for c in prefix:
    let index = ord(c) - ord('a')

    if current.children[index] == nil:
      return false

    current = current.children[index]

  return true


proc findStartsWith*(t: Trie, prefix: string): seq[string] =
  var current = t.root

  for c in prefix:
    let index = ord(c) - ord('a')

    if current.children[index] == nil:
      return @[]

    current = current.children[index]

  result = @[]
  var currentPrefix = prefix
  collectWords(current, currentPrefix, result)

proc collectWords(node: TrieNode, current: var string, results: var seq[string]) =
  if node.isEndOfWord:
    results.add(current)

  for i in 0..25:
    if node.children[i] != nil:
      let c = char(ord('a') + i)
      current.add(c)
      collectWords(node.children[i], current, results)
      current.setLen(current.len - 1)


proc `$`*(t: Trie): string =
  var sb = ""
  buildString(t.root, sb, "<root>", "", "")
  result = sb

proc buildString(node: TrieNode, sb: var string, label: string, prefix: string, childPrefix: string) =
  sb.add(prefix & label & (if node.isEndOfWord: "*" else: "") & "\n")

  for i, child in node.children:
    if child != nil:
      let childLabel = $char(ord('a') + i)
      buildString(child, sb, childLabel, childPrefix & "├── ", childPrefix & "│   ")

File: programming-languages/java/trie/src/main/java/com/ac/Trie.java

package com.ac;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Trie {

    static class TrieNode {
        public TrieNode[] children;
        public boolean isEndOfWord;
        public char code; // only for debug purposes

        public TrieNode(char code) {
            children = new TrieNode[26];
            isEndOfWord = false;
            this.code = code;
        }

        @Override
        public String toString() {
            return "" + code;
        }
    }

    private final TrieNode root;

    public Trie() {
        root = new TrieNode('0');
    }

    public TrieNode getRoot() {
        return root;
    }

    public void insert(String word) {
        TrieNode currentNode = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (currentNode.children[index] == null) {
                currentNode.children[index] = new TrieNode(c);
            }
            currentNode = currentNode.children[index];
        }
        currentNode.isEndOfWord = true;
    }

    public boolean search(String word) {
        TrieNode currentNode = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (currentNode.children[index] == null) {
                return false;
            }
            currentNode = currentNode.children[index];
        }
        return currentNode.isEndOfWord;
    }

    public boolean startsWith(String prefix) {
        TrieNode current = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            int index = c - 'a';
            if (current.children[index] == null) {
                return false;
            }

            current = current.children[index];
        }
        return true;
    }

    public List<String> findStartsWith(String prefix) {
        TrieNode current = root;
        for (char c : prefix.toCharArray()) {
            int index = c - 'a';
            if (current.children[index] == null) {
                return Collections.emptyList();
            }
            current = current.children[index];
        }
        List<String> results = new ArrayList<>();
        collectWords(current, new StringBuilder(prefix), results);
        return results;
    }

    private void collectWords(TrieNode node, StringBuilder current, List<String> results) {
        if (node.isEndOfWord) {
            results.add(current.toString());
        }
        for (int i = 0; i < 26; i++) {
            if (node.children[i] != null) {
                current.append((char) ('a' + i));
                collectWords(node.children[i], current, results);
                current.deleteCharAt(current.length() - 1);
            }
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        buildString(root, sb, "", "");
        return sb.toString();
    }

    private void buildString(TrieNode node, StringBuilder sb, String prefix, String childPrefix) {
        sb.append(prefix).append(node.code).append(node.isEndOfWord ? "*" : "").append("\n");
        for (int k = 0; k < node.children.length; k++) {
            var child = node.children[k];
            if (child != null) {
                buildString(child, sb, childPrefix + "├── ", childPrefix + "│   ");
            }
        }
    }

}
